rule-engine/rule-engine-components/src/test/java/org/thingsboard/rule/engine/delay/TbMsgDelayNodeTest.java

```diff
@@ -0,0 +1,319 @@
+    }
+
+    @Test
+    public void givenInvalidTimeUnit_whenOnMsg_thenThrowsException() throws TbNodeException {
+        config.setTimeUnit("sec");
+        node.init(ctxMock, new TbNodeConfiguration(JacksonUtil.valueToTree(config)));
+
+        var msg = TbMsg.newMsg(TbMsgType.POST_TELEMETRY_REQUEST, DEVICE_ID, TbMsgMetaData.EMPTY, TbMsg.EMPTY_JSON_OBJECT);
+        assertThatThrownBy(() -> node.onMsg(ctxMock, msg))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessage("Invalid value for period time unit : sec");
+    }
+
+    @Test
+    public void givenMaxLimitOfPendingMsgsReached_whenOnMsg_thenTellFailure() throws TbNodeException {
+        config.setMaxPendingMsgs(1);
+
+        node.init(ctxMock, new TbNodeConfiguration(JacksonUtil.valueToTree(config)));
+        var msg = TbMsg.newMsg(TbMsgType.POST_TELEMETRY_REQUEST, DEVICE_ID, TbMsgMetaData.EMPTY, TbMsg.EMPTY_JSON_OBJECT);
+        for (int i = 0; i < 2; i++) {
+            node.onMsg(ctxMock, msg);
+        }
+
+        ArgumentCaptor<Throwable> throwable = ArgumentCaptor.forClass(Throwable.class);
+        then(ctxMock).should().tellFailure(eq(msg), throwable.capture());
+        assertThat(throwable.getValue()).isInstanceOf(RuntimeException.class).hasMessage("Max limit of pending messages reached!");
+    }
+
+    @Test
+    public void verifyDestroyMethod() {
+        var msg = TbMsg.newMsg(TbMsgType.POST_TELEMETRY_REQUEST, DEVICE_ID, TbMsgMetaData.EMPTY, TbMsg.EMPTY_JSON_OBJECT);
+        var pendingMsgs = new ConcurrentHashMap<>();
+        pendingMsgs.put(UUID.fromString("321f0301-9bed-4e7d-b92f-a978f53ec5d6"), msg);
+        ReflectionTestUtils.setField(node, "pendingMsgs", pendingMsgs);
+        var actualPendingMsgs = (Map<UUID, TbMsg>) ReflectionTestUtils.getField(node, "pendingMsgs");
+        assertThat(actualPendingMsgs).isEqualTo(pendingMsgs);
+
+        node.destroy();
+
+        assertThat(actualPendingMsgs).isEmpty();
+    }
+
+    private static Stream<Arguments> givenFromVersionAndConfig_whenUpgrade_thenVerifyHasChangesAndConfig() {
+        return Stream.of(
+                // config for version 1 with upgrade from version 0 (useMetadataPeriodInSecondsPatterns does not exist)
+                Arguments.of(0,
+                        """
+                                    {
+                                        "periodInSeconds": 13,
+                                        "maxPendingMsgs": 1000,
+                                        "periodInSecondsPattern": "17"
+                                    }
+                                """,
+                        true,
+                        """
+                                    {
+                                        "period": "60",
+                                        "timeUnit": "SECONDS",
+                                        "maxPendingMsgs": 1000
+                                    }
+                                """
```

Comment:
why period was changed from 13 seconds to 60 seconds?

Reply:
> **If there is no useMetadataPeriodInSecondsPatterns then set period to "60"**

This feature didn't exist at the time the rule node was created. So potentially there might be cases where this parameter doesn't exist in the rule node configuration. Non-existing boolean parameters are treated as false during the parse of json config to the configuration class..